### ISSUE TITLE
Path::components("///a/b").as_path() failing to normalize multiple slashes for root dir on Unix

### ISSUE THREAD

**Issue by programmerjake** [rust-lang/rust#83912]
I'd expect the path to be normalized to `"/a/b"` since that is a substring of `"///a/b"`.

On some Unix implementations a leading `//` is special, but the Unix and POSIX specs state that a leading `///` is equivalent to just one `/`. On Linux, `//` is also the same as `/`, though it has been proposed for that to change.
[Stack Exchange: On what systems is //foo/bar different from /foo/bar?](https://unix.stackexchange.com/questions/256497/on-what-systems-is-foo-bar-different-from-foo-bar)
```rust
use std::path::Path;

fn main() {
    let expected = "/a/b";
    let actual = Path::new("///a/b").components().as_path().to_str().unwrap();
    assert_eq!(expected, actual);
}
```

([Playground](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=192f448ca2b62a3bf0bbb135addcda25))

Errors:

```
   Compiling playground v0.0.1 (/playground)
    Finished release [optimized] target(s) in 3.78s
     Running `target/release/playground`
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `"/a/b"`,
 right: `"///a/b"`', src/main.rs:6:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

**Comment by the8472** [comment]
`components()` 

> When **parsing** the path, there is a small amount of normalization

`.as_path()`

> Extracts a **slice** corresponding to the portion of the path remaining for iteration.


Since slices are contiguous they cannot be the result of parsing and thus don't include normalization.

**Comment by programmerjake** [comment]
@the8472
> Since slices are contiguous they cannot be the result of parsing and thus don't include normalization.

That's not correct, normalization can be done as long as the result is either a substring of the input or a static string.

For example, `"a/./b/./."` *is* normalized to `"a/./b"` since `components()` can just reduce the length to removing the trailing `"/./."`:
https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=0730bdbe3f6dc1dc17b46b7bd8c5bfdb

**Comment by the8472** [comment]
Sure, but in the general case it *cannot* guarantee normalization, therefore the question is where does the expectation come from? It's not part of the API.

**Comment by programmerjake** [comment]
The docs for `Path::components` state that:
> Repeated separators are ignored, so `a/b` and `a//b` both have `a` and `b` as components.

I'd argue that all but the first slash in `///a//b/.` (switching to a better example) are separators, so the Path should be normalized such that only the necessary separators show up: `/a//b` -- the initial slash is needed to denote the root dir, and the slashes between `a` and `b` are necessary to avoid allocating. All other slashes are trimmed from the underlying `OsStr` -- normalization to a substring.

**Comment by programmerjake** [comment]
alternatively, if separator is interpreted to mean the `/` character specifically (on Unix), then that makes an even better argument for normalizing `///a//b/.` to `/a//b`

**Comment by the8472** [comment]
My argument is that the normalization described in `Path::component` logically cannot be part of the `Components::as_path` contract because separators in the middle cannot be normalized if the returned value is a slice. Therefore we can only rely on what `as_path` itself promises, which doesn't mention any normalization.

**Comment by eggyal** [comment]
For the sake of absolute clarity: what the8472 is saying is that `components()` already normalizes repeated path delimiters (wherever they arise, prefix or not) and does not emit any empty component here; your problem is that `as_path` then collapses what remains of the components, effectively discarding any normalization that the iterator would have performed thereafter.  If you want to construct a new, normalized path from the components, `collect` it into a `PathBuf`:

```rust
use std::path::{Path, PathBuf};

fn main() {
    let expected = "/a/b";
    let actual = Path::new("///a/b").components().collect::<PathBuf>();
    let actual = actual.to_str().unwrap();
    assert_eq!(expected, actual);
}
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=10673d4cdbe5c222e456c057b0763f34)

**Comment by programmerjake** [comment]
I think everyone can agree that `.components().collect()` does a full normalization (to the extent specified by `components`'s docs). The issue here is that `.components().as_path()` does less normalization than expected even when considering the constraints of not allocating.

**Comment by eggyal** [comment]
Apologies, clearly I'd missed the point here!!

**Comment by programmerjake** [comment]
> Maybe you're really looking to [`canonicalize`](https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.canonicalize) instead?

nope, I'm writing a reimplementation of `Path` for use in a virtualized/sandboxed Linux filesystem (even if the host OS isn't Unix), I definitely don't want it to access the host's filesystem.

**Comment by programmerjake** [comment]
> Apologies, clearly I'd missed the point here!!

no problem! If I weren't writing a virtual filesystem, `canonicalize` or `components().collect()` would definitely be useful!

**Comment by the8472** [comment]
> The issue here is that .components().as_path() does less normalization than expected even when considering the constraints of not allocating.

What I am trying to say is that one shouldn't expect anything from it since the `components()` contract is clearly not applicable.  `as_path()` itself on the other hand only promises to return the tail of whatever remains in the iterator. And since iterators are lazy (per `core::iter` module documentation) one can't expect it to do any work to do additional parsing or normalization.

So the default expectation should be to just get a slice of whatever remains in the iterator (as the documentation says), which may contain some non-normalized path elements. That it currently happens to strip things off the tail is something going beyond its contract, but that shouldn't inform further expectations, on the opposite, should code defensively because this current behavior is not guaranteed and could thus go away at any point in the future.

Imo there's no failure here, the methods are within specification.

**Comment by eggyal** [comment]
> Imo there's no failure here, the methods are within specification.

So feature request then?

@rustbot label: +T-libs +C-feature-request

**Comment by marshray** [comment]
I will politely disagree with the proposition that

> one shouldn't expect anything from `.as_path()` since the components() contract is clearly not applicable.

That was certainly not clear to me. Instead, what it did was set off my "well how could that possibly work, could this be a bug?" detector and, as a result, I spent time learning about this today that I might have instead spent on other things.

Where does one usually look for information first: the doc for the struct whose only purpose in life is to `impl Iterator`, or the collection method that returns it?

The doc for [Components](https://doc.rust-lang.org/stable/std/path/struct.Components.html) says explicitly "This `struct` is created by the `components()` method on `Path`. See its documentation for more." and [Path::components()](https://doc.rust-lang.org/stable/std/path/struct.Path.html#method.components) is the *only* way to obtain a `Components`.

`Path::components()` specifically states that "Occurrences of `.` are normalized away, except if they are at the beginning of the path". Yet, if one calls `p.components()`[`.as_path()`](https://doc.rust-lang.org/stable/std/path/struct.Components.html#method.as_path) it only normalizes away the `.`s at the *end* of the path.

> as_path() itself on the other hand only promises to return the tail of whatever remains in the iterator.

What does "remaining in an iterator" mean if not "the sequence of things that would be produced by successive calls to `.next()`"?

Specifically, the doc says: "Extracts a slice corresponding to the portion of the path remaining for iteration."

But I think this probably only makes sense to someone who already understands the implementation model. I.e., you're looking at the source and seeing that `Components` is really just a `&'a [u8]` plus some other state to handle various corner cases.

My request would be a simple clarification to the documentation. For example, `.as_path()` could be updated to something like:
* "Returns a ref to a sub-slice of the `Path` corresponding to the portion remaining for iteration. Avoids allocation. As a consequence, the "small amount of normalization" described in `Path.components()` will have only been applied to the beginning and end of the returned section. Any interior `.` components will still be present, even if they would have been skipped during iteration. Consider using `iter.collect::<PathBuf>()` if that normalization is needed."

One might even consider a new method, e.g. `Components<'a>::to_path(&self) -> Cow<'a, Path>`, which allocates only in the (less common) case where the path contains an interior `.`.

I'd be happy to submit a PR if that would be useful.

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5d44cea7ce3a5a8adab4849d15e48600) (Stable 1.88)

```
    let p: &Path = Path::new("a/./b/./c");
    
    println!("path                 : {p:?}");
    println!("components           : {:?}", p.components().collect::<Vec<_>>());

    let mut pc = p.components();

    println!("components.as_path() : {:?}", pc.as_path());
    println!("components.collect   : {:?}", pc.clone().collect::<PathBuf>());
    println!("next_back()"); let _ = pc.next_back();
    println!("components.as_path() : {:?}", pc.as_path());
    println!("components.collect   : {:?}", pc.clone().collect::<PathBuf>());
    println!("next_back()"); let _ = pc.next_back();
    println!("components.as_path() : {:?}", pc.as_path());
    println!("components.collect   : {:?}", pc.clone().collect::<PathBuf>());
```

```
path                 : "a/./b/./c"
components           : [Normal("a"), Normal("b"), Normal("c")]
components.as_path() : "a/./b/./c"
components.collect   : "a/b/c"
next_back()
components.as_path() : "a/./b"
components.collect   : "a/b"
next_back()
components.as_path() : "a"
components.collect   : "a"
```
